examples/platforms/efr32mg12/radio.c

```diff
@@ -0,0 +1,875 @@
+{
+    RAIL_Status_t              status;
+    RAIL_TxPowerCurvesConfig_t txPowerCurvesConfig = {curves24Hp, curvesSg, curves24Lp, piecewiseSegments};
+
+    status = RAIL_InitTxPowerCurves(&txPowerCurvesConfig);
+    assert(status == RAIL_STATUS_NO_ERROR);
+
+    status = RAIL_SetTxPowerDbm(gRailHandle, ((RAIL_TxPower_t)aPowerDbm) * 10);
+    assert(status == RAIL_STATUS_NO_ERROR);
+}
+
+static efr32BandConfig *efr32RadioGetBandConfig(uint8_t aChannel)
+{
+    efr32BandConfig *config = NULL;
+
+    for (uint8_t i = 0; i < EFR32_NUM_BAND_CONFIGS; i++)
+    {
+        if ((sBandConfigs[i].mChannelMin <= aChannel) && (aChannel <= sBandConfigs[i].mChannelMax))
+        {
+            config = &sBandConfigs[i];
+            break;
+        }
+    }
+
+    return config;
+}
+
+static void efr32ConfigInit(void (*aEventCallback)(RAIL_Handle_t railHandle, RAIL_Events_t events))
+{
+    sCommonConfig.mRailConfig.eventsCallback = aEventCallback;
+    sCommonConfig.mRailConfig.protocol       = NULL; // only used by Bluetooth stack
+    sCommonConfig.mRailConfig.scheduler      = NULL; // only needed for DMP
+
+    uint8_t index = 0;
+
+#if RADIO_CONFIG_2P4GHZ_OQPSK_SUPPORT
+    sBandConfigs[index].mChannelConfig = NULL;
+    sBandConfigs[index].mChannelMin    = OT_RADIO_2P4GHZ_OQPSK_CHANNEL_MIN;
+    sBandConfigs[index].mChannelMax    = OT_RADIO_2P4GHZ_OQPSK_CHANNEL_MAX;
+
+    index++;
+#endif
+
+#if RADIO_CONFIG_915MHZ_OQPSK_SUPPORT
+    sBandConfigs[index].mChannelConfig = channelConfigs[0];
+    sBandConfigs[index].mChannelMin    = OT_RADIO_915MHZ_OQPSK_CHANNEL_MIN;
+    sBandConfigs[index].mChannelMax    = OT_RADIO_915MHZ_OQPSK_CHANNEL_MAX;
+#endif
+    gRailHandle = efr32RailInit(&sCommonConfig);
+    assert(gRailHandle != NULL);
+    efr32RailConfigLoad(&(sBandConfigs[0]));
+}
+
+void efr32RadioInit(void)
+{
+    RAIL_Status_t status;
+
+    // check if RAIL_TX_FIFO_SIZE is power of two..
+    assert((RAIL_TX_FIFO_SIZE & (RAIL_TX_FIFO_SIZE - 1)) == 0);
+
+    // check the limits of the RAIL_TX_FIFO_SIZE.
+    assert((RAIL_TX_FIFO_SIZE >= 64) || (RAIL_TX_FIFO_SIZE <= 4096));
+
+    efr32ConfigInit(RAILCb_Generic);
+
+    CMU_ClockEnable(cmuClock_PRS, true);
+    RTCDRV_Init();
+    status = RAIL_ConfigSleep(gRailHandle, RAIL_SLEEP_CONFIG_TIMERSYNC_ENABLED);
+    assert(status == RAIL_STATUS_NO_ERROR);
+
+    sReceiveFrame.mLength  = 0;
+    sReceiveFrame.mPsdu    = sReceivePsdu;
+    sTransmitFrame.mLength = 0;
+    sTransmitFrame.mPsdu   = sTransmitPsdu;
+
+    sCurrentBandConfig = efr32RadioGetBandConfig(OPENTHREAD_CONFIG_DEFAULT_CHANNEL);
+    assert(sCurrentBandConfig != NULL);
+
+    efr32RadioSetTxPower(sTxPowerDbm);
+
+    sEnergyScanStatus = ENERGY_SCAN_STATUS_IDLE;
+
+    otLogInfoPlat("Initialized", NULL);
+}
+
+void efr32RadioDeinit(void)
+{
+    RAIL_Status_t status;
+
+    status = RAIL_ConfigEvents(gRailHandle, RAIL_EVENTS_ALL, 0);
+    assert(status == RAIL_STATUS_NO_ERROR);
+    RAIL_Idle(gRailHandle, RAIL_IDLE_ABORT, true);
```

Comment:
Would idling the radio prior to turning off the events be safer?  What happens if the radio is aborted in the middle of an event?  e.g. sTransmitBusy is cleared by the RAIL_EVENTS_TX_COMPLETION event being processed, but if the events are disabled prior to the radio being idled, is it possible that this flag could get stuck set?

Reply:
I helped @hemanth-silabs with the RAIL drivers, so I reply here. I'm not an expert on openthread, but I do have a lot of experience in RAIL.

Since the event handler is almost always running in interrupt context (including TX_COMPLETION), I don't think it's possible to abort an ongoing event. However, calling RAIL_IDLE_ABORT can cause RAIL_EVENT_TX_ABORTED or RAIL_EVENT_RX_PACKET_ABORTED, if called during frame tx/rx, and we don't want to handle that event.

Reply:
The RAIL documentation appears to say that it is possible to abort an ongoing event: "It also aborts any receive, transmit, or scheduled events in progress" which is what gives me some pause here.  What potential issues do you see that could arise in handling the ABORTED events?  The handler already handles RAIL_EVENT_TX_ABORTED by clearing the sTransmitBusy flag and setting the sTransmitError to OT_ERROR_ABORT.  The RAIL_EVENT_RX_ABORTED could perhaps just be dropped in the callback handler since nothing would need to be done.

Reply:
Well, the sentence you quote uses unfortunate wording. I think scheduled tasks would be better, since it's not about the event handler, it's about ongoing rx/tx (which can be aborted, and will result ABORTED event)

The idea behind this implementation is that we want to stop the radio with efr32RadioDeinit, and we don't want any further notification from the driver. If we would swap the order, the driver could call `otPlatRadioTxDone` or similar Done functions.

If that's not the goal of this function, we can swap them.

Reply:
I'm okay with the idea of not handling the event(s) when deinitializing the radio.  The concern I have is with stale data being left in sTransmitError and sTransmitBusy, in the event that the radio is re-initialized at a later point in time.  Can we clear those variables in efr32RadioInit()?

Reply:
Variables are now been cleared in efr32RadioInit in the later commit 657bf55e7.